tinygrad/tensor.py

```diff
@@ -348,34 +346,35 @@ def __getitem__(self, indices: Union[int, slice, Tensor, None, List, Tuple]) ->
     if float in type_dim: raise IndexError("float type is not valid index")
     if any(isinstance(i, slice) and i.step == 0 for i in indices): raise ValueError('slice step cannot be 0')
     if num_slices > len(self.shape): raise IndexError(f"too many indices for tensor of dimension {len(self.shape)}")
-    for dim in type_dim[int]:
-      if indices_filtered[dim] >= self.shape[dim] or indices_filtered[dim] < -self.shape[dim]:
-        raise IndexError(f"index {indices_filtered[dim]} is out of bounds for dimension {dim} with size {self.shape[dim]}")
-
-    # normalize! indices -> start, stop, strides
-    # TODO: this line is completely unreadable
-    start, stop, strides = zip(*y) if (y := [i.indices(sh) if isinstance(i, slice) else slice(normalized:= i if i != -1 else sh-1, normalized+1, 1).indices(sh) if isinstance(i, int) else (0, sh, 1) for i, sh in zip(indices_filtered, self.shape)]) else ((), (), ()) # type: ignore[arg-type]   # noqa: E501
 
     # 2. basic indexing (no copy)
-    # apply slices and flip where strides are negative
-    new_slice = tuple(((0, 0) if e < s else (s, e)) if st > 0 else ((0, 0) if e > s else (e+1, s+1)) for s, e, st in zip(start, stop, strides))
-    sliced_tensor = self.shrink(new_slice).flip(axis=[i for i, s in enumerate(strides) if s < 0])
-    new_shape = list(sliced_tensor.shape)
-
-    # add strides by pad -> reshape -> shrink
-    if any(abs(s) != 1 for s in strides):
-      strides = tuple(abs(s) for s in strides)
-      padded_tensor = sliced_tensor.pad(tuple((0, s-(dim_sz % s) if dim_sz % s != 0 else 0) for s, dim_sz in zip(strides, sliced_tensor.shape)))
-      reshaped_tensor = padded_tensor.reshape(flatten([sh // s, s] for sh, s in zip(padded_tensor.shape, strides)))
-      new_shape = list(reshaped_tensor.shape[::2])
-      sliced_tensor = reshaped_tensor.shrink(tuple(flatten(((0, sh), (0, 1)) for sh in new_shape)))
-
-    # inject dim=1 for None and collapse dim for int
+    if indexed := type_dim[int] or type_dim[slice]:
+      # turn indices in indices_filtered to Tuple[shrink_arg, strides]
+      for dim in type_dim[int]:
+        if (i := indices_filtered[dim]) >= self.shape[dim] or i < -self.shape[dim]:
+          raise IndexError(f"index {i} is out of bounds for dimension {dim} with size {self.shape[dim]}")
+        indices_filtered[dim] = ((i,i+1),1) if i >= 0 else ((self.shape[dim]+i, self.shape[dim]+i+1), 1)
+      for dim in type_dim[slice]:
+        s, e, st = indices_filtered[dim].indices(self.shape[dim])
+        indices_filtered[dim] = (((0, 0) if e < s else (s, e)) if st > 0 else ((0, 0) if e > s else (e+1, s+1)), st)
```

Comment:
multiple conditions like this is very hard to read, think about what's common in them and can be factor out